File: packages/js/core/src/genai/generator.ts
import { Tensor } from '../ir/tensor.js';
import { State } from './state.js';
import { GeneratorParams } from './types.js';

/**
 * Base Generator class for stateful decoding.
 */
export abstract class Generator {
  protected state: State;
  protected params: GeneratorParams;

  constructor(state: State, params: GeneratorParams) {
    this.state = state;
    this.params = params;
  }

  /**
   * Compute logits for the current state.
   */
  abstract computeLogits(inputIds: Tensor): Promise<Tensor>;

  /**
   * Compute logits synchronously (if supported).
   */
  abstract computeLogitsSync(inputIds: Tensor): Tensor;

  /**
   * Process pre-fill phase.
   */
  abstract prefill(promptIds: Tensor): Promise<Tensor>;

  /**
   * Process a single decoding step.
   */
  abstract decodeStep(tokenId: number): Promise<Tensor>;

  /**
   * High-level generation API. Yields tokens as they are generated.
   */
  async *generate(promptIds: Tensor): AsyncGenerator<number, void, unknown> {
    let currentTokens = 0;
    const maxTokens =
      this.params.maxNewTokens ??
      this.params.maxLength - (promptIds.shape[promptIds.shape.length - 1] as number);

    let logits = await this.prefill(promptIds);
    let nextToken = this.sample(logits);

    yield nextToken;
    currentTokens++;

    while (currentTokens < maxTokens) {
      // Check early stopping
      if (this.params.abortSignal?.aborted) {
        break;
      }
      if (this.params.earlyStopping && this.isEos(nextToken)) {
        break;
      }

      logits = await this.decodeStep(nextToken);
      nextToken = this.sample(logits);

      yield nextToken;
      currentTokens++;
    }
  }

  /**
   * Sample the next token from logits.
   */
  protected sample(logits: Tensor): number {
    // Basic greedy search implementation for now.
    // Needs proper logit processing pipeline later.
    if (logits.data instanceof Float32Array || logits.data instanceof Float64Array) {
      const data = logits.data;
      const vocabSize = logits.shape[logits.shape.length - 1] as number;
      // Get the last logits
      const offset = data.length - vocabSize;
      let maxVal = -Infinity;
      let maxIdx = -1;
      for (let i = 0; i < vocabSize; i++) {
        if (data[offset + i]! > maxVal) {
          maxVal = data[offset + i]!;
          maxIdx = i;
        }
      }
      return maxIdx;
    }
    throw new Error('Unsupported logit data type for sampling.');
  }

  protected isEos(tokenId: number): boolean {
    // Implement proper EOS checking logic with ModelParams
    return false; // placeholder
  }
}

File: packages/python/onnx9000-core/src/onnx9000/genai/generator.py
"""Base implementation of the stateful generation engine for the onnx9000 ecosystem.

This module provides the core Generator class which orchestrates the prefill and
decoding phases of autoregressive model generation.
"""

import struct
from collections.abc import AsyncIterator

from ..core.ir import Tensor
from .state import State
from .types import GeneratorParams


class Generator:
    """Orchestrates the autoregressive generation process for a stateful model.

    The Generator manages the transition between the initial prefill phase
    and the subsequent incremental decoding steps, utilizing a state object
    to maintain KV caches and other persistent information.
    """

    def __init__(self, state: State, params: GeneratorParams):
        """Initialize the Generator with a specific execution state and generation parameters.

        Args:
            state: The stateful context for the model, including KV caches.
            params: Configuration for the generation process (e.g., max_tokens, sampling).
        """
        self.state = state
        self.params = params

    async def compute_logits(self, input_ids: Tensor) -> Tensor:
        """Perform forward pass to compute logits for the given input identifiers.

        Args:
            input_ids: A tensor of token IDs to process.

        Returns:
            A Tensor containing the output logits for the last token in each sequence.
        """
        return None

    def compute_logits_sync(self, input_ids: Tensor) -> Tensor:
        """Compute logits synchronously.

        This is a blocking version of compute_logits for backends that do not
        support asynchronous execution.

        Args:
            input_ids: A tensor of token IDs to process.

        Returns:
            A Tensor containing the output logits.
        """
        return None

    async def prefill(self, prompt_ids: Tensor) -> Tensor:
        """Process the initial prompt to populate the KV cache and get first logits.

        Args:
            prompt_ids: The initial prompt token identifiers.

        Returns:
            The logits produced by processing the full prompt.
        """
        return None

    async def decode_step(self, token_id: int) -> Tensor:
        """Process a single new token and update the generation state.

        Args:
            token_id: The ID of the most recently generated or sampled token.

        Returns:
            The logits for the next token prediction.
        """
        return Tensor(name="dummy", shape=[], data=None)

    async def generate(self, prompt_ids: Tensor) -> AsyncIterator[int]:
        """Generate a sequence of tokens from a given prompt.

        This is the primary entry point for generation. It yields tokens one-by-one
        as they are produced by the model.

        Args:
            prompt_ids: The starting prompt for generation.

        Yields:
            Generated token IDs until an EOS condition is met or limits are reached.
        """
        current_tokens = 0
        prompt_len = prompt_ids.shape[-1] if prompt_ids.shape else 0

        max_tokens = self.params.max_new_tokens
        if max_tokens is None:
            max_tokens = self.params.max_length - prompt_len

        logits = await self.prefill(prompt_ids)
        next_token = self.sample(logits)

        yield next_token
        current_tokens += 1

        while current_tokens < max_tokens:
            if self.params.abort_signal:
                break
            if self.params.early_stopping and self.is_eos(next_token):
                break

            logits = await self.decode_step(next_token)
            next_token = self.sample(logits)

            yield next_token
            current_tokens += 1

    def sample(self, logits: Tensor) -> int:
        """Select the next token ID from the predicted probability distribution.

        Currently implements greedy sampling (argmax).

        Args:
            logits: The output logits from the model's last layer.

        Returns:
            The ID of the selected next token.
        """
        # Simple argmax greedy search
        if logits.data is None:
            return 0

        vocab_size = logits.shape[-1] if logits.shape else 1
        itemsize = logits.dtype.itemsize if hasattr(logits.dtype, "itemsize") else 4

        # Assuming float32 for simplicity
        offset = len(logits.data) - vocab_size * itemsize
        if offset < 0:
            return 0

        max_val = float("-inf")
        max_idx = -1

        for i in range(vocab_size):
            start = offset + i * itemsize
            if start + itemsize <= len(logits.data):
                val = struct.unpack("<f", logits.data[start : start + itemsize])[0]
                if val > max_val:
                    max_val = val
                    max_idx = i

        return max_idx if max_idx != -1 else 0

    def is_eos(self, token_id: int) -> bool:
        """Check if the given token ID represents an End-Of-Sequence marker.

        Args:
            token_id: The token ID to check.

        Returns:
            True if the token is an EOS token, False otherwise.
        """
        return False
